source/snapshot/fetch_worker.go

```diff
@@ -0,0 +1,393 @@
+	}
+	defer closeCursor()
+
+	var nfetched int64
+
+	for {
+		n, err := f.fetch(ctx, tx)
+		if err != nil {
+			return fmt.Errorf("failed to fetch results: %w", err)
+		}
+
+		if n == 0 { // end of cursor
+			break
+		}
+
+		nfetched += int64(n)
+
+		sdk.Logger(ctx).Info().
+			Int64("rows", nfetched).
+			Str("table", f.conf.Table).
+			Dur("elapsed", time.Since(start)).
+			Msg("fetching rows")
+	}
+
+	sdk.Logger(ctx).Info().
+		Dur("elapsed", time.Since(start)).
+		Str("table", f.conf.Table).
+		Msgf("%q snapshot completed", f.conf.Table)
+
+	return nil
+}
+
+func (f *FetchWorker) createCursor(ctx context.Context, tx pgx.Tx) (func(), error) {
+	cursorSQL := fmt.Sprintf("DECLARE %s CURSOR FOR (SELECT * FROM %s WHERE %s > %d AND %s <= %d ORDER BY %s)",
+		f.cursorName,
+		f.conf.Table,
+		f.conf.Key,
+		f.lastRead,
+		f.conf.Key,
+		f.snapshotEnd,
+		f.conf.Key,
+	)
+
+	if _, err := tx.Exec(ctx, cursorSQL); err != nil {
+		return nil, err
+	}
+
+	return func() {
+		// N.B. The cursor will automatically close when the TX is done.
+		if _, err := tx.Exec(ctx, "CLOSE "+f.cursorName); err != nil {
+			sdk.Logger(ctx).Warn().
+				Err(err).
+				Msgf("unexpected error when closing cursor %q", f.cursorName)
+		}
+	}, nil
+}
+
+func (f *FetchWorker) updateFetchLimit(ctx context.Context, tx pgx.Tx) error {
+	if f.snapshotEnd > 0 {
+		return nil
+	}
+
+	if err := tx.QueryRow(
+		ctx,
+		fmt.Sprintf("SELECT max(%s) FROM %s", f.conf.Key, f.conf.Table),
+	).Scan(&f.snapshotEnd); err != nil {
+		return fmt.Errorf("failed to query max on %q.%q: %w", f.conf.Table, f.conf.Key, err)
+	}
+
+	return nil
+}
+
+func (f *FetchWorker) fetch(ctx context.Context, tx pgx.Tx) (int, error) {
+	rows, err := tx.Query(ctx, fmt.Sprintf("FETCH %d FROM %s", f.conf.FetchSize, f.cursorName))
+	if err != nil {
+		return 0, fmt.Errorf("failed to fetch rows: %w", err)
+	}
+	defer rows.Close()
+
+	var fields []string
+	for _, f := range rows.FieldDescriptions() {
+		fields = append(fields, f.Name)
+	}
+
+	var nread int
+
+	for rows.Next() {
+		values, err := rows.Values()
+		if err != nil {
+			return 0, fmt.Errorf("failed to get values")
+		}
+
+		if err := f.send(
+			ctx,
+			f.buildRecord(fields, values),
+		); err != nil {
+			return nread, fmt.Errorf("failed to send record")
+		}
+
+		nread++
+	}
+
+	return nread, nil
+}
+
+func (f *FetchWorker) send(ctx context.Context, r sdk.Record) error {
+	select {
+	case <-ctx.Done():
+		return fmt.Errorf("fetcher send ctx: %w", ctx.Err())
+	case f.out <- r:
+		return nil
+	}
+}
+
+func (f *FetchWorker) buildRecord(fields []string, values []any) sdk.Record {
+	payload := make(sdk.StructuredData)
+
+	for i, name := range fields {
+		switch t := values[i].(type) {
+		case time.Time: // type not supported in sdk.Record
+			payload[name] = t.UTC().String()
+		default:
+			payload[name] = t
+		}
+	}
+
+	// always coerce snapshot position to bigint, pk may be any type of integer.
+	lastRead := keyInt64(payload[f.conf.Key])
```

Comment:
Are we confident this will never panic? Otherwise I'd rather return an error.

Reply:
Well, actually, it probably will, since the types we are allow are also real/double and numeric. 

I think the ideal situation is to drop all non-integers, and then this will be fine. Having pk to be numeric or real is something odd...

Reply:
I changed this to return an error so we at least don't panic right now, in case we encounter this.